### PR TITLE
Migrate vmtest to modular actions in libbpf/ci

### PR DESCRIPTION
**Pull request by thefallentree**
1. break debian/ubuntu build to separate workflow file, since they don't share anything with kernel selftests
2. breakdown monolithic vmtest action into documented modular steps
3. move some of the common steps to libbpf/ci ,currently saved at thefallentree branch

### CODE REVIEW

**Comment by thefallentree** [comment]
I think  the s390 failure needs a update to the runner, but otherwise everything passes.

**Comment by anakryiko** [comment]
Looks pretty good, thank you. Some steps seem a bit too granular (e.g., download sources + get vmlinux image for !LATEST + compile kernel for LATEST seems like should be just one command which maybe takes a few input arguments (e.g., kernel config, maybe source code location)). But that we can iterate upon.

But please make sure that build log is clean, right now it has lots of non-collapsed stuff, which compared to what we had previously is much harder to work with.

Screen Shot 2021-12-03 at 11 44 22 AM

vs 

Screen Shot 2021-12-03 at 11 47 05 AM

**Comment by thefallentree** [comment]
PTAL, I've cleaned up the output log and fixed s390x/x86_64 LATEST build

**Comment by anakryiko** [comment]
> PTAL, I've cleaned up the output log

not really?.. first few steps are still messed up:

Screen Shot 2021-12-06 at 6 27 31 PM

**Review by anakryiko — COMMENTED** [review]
damn, I forgot to post my comment from the previous round, sorry about that :(

File: .github/actions/build-selftests/build_selftests.sh [hunk 1]
```diff
@@ -2,41 +2,41 @@
...
+cd ${REPO_ROOT}/${REPO_PATH}
 make \
 	CLANG=clang-${LLVM_VER} \
 	LLC=llc-${LLVM_VER} \
 	LLVM_STRIP=llvm-strip-${LLVM_VER} \
 	VMLINUX_BTF="${VMLINUX_BTF}" \
 	VMLINUX_H=${VMLINUX_H} \
 	-C "${REPO_ROOT}/${REPO_PATH}/tools/testing/selftests/bpf" \
-	-j $((4*$(nproc))) >/dev/null
+	-j $((4*$(nproc)))
```
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/build-selftests/build_selftests.sh:33]: piping to /dev/null will make it harder to debug
while the action is running, without constant output it won't show anything to tell you what step it is stuck on.  
the ouput is folded anyway, so it wasn't a issue.

File: .github/actions/vmtest/action.yml [hunk 2]
```diff
@@ -5,31 +5,72 @@ inputs:
...
+        make olddefconfig && make prepare
+        cd -
+    # 2. if kernel == LATEST, build kernel image from tree
+    - name: Build kernel image
+      if: ${{ inputs.kernel == 'LATEST' }}
       shell: bash
+      run: |
+        echo "::group::Build Kernel Image"
+        cd .kernel
+        make -j $((4*$(nproc))) all
```
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/vmtest/action.yml:52]: same as above, the output is folded once it's finished, so it isn't a issue.

File: .github/actions/vmtest/action.yml [hunk 3]
```diff
@@ -5,31 +5,72 @@ inputs:
...
+      run: |
+        cat CHECKPOINT-COMMIT
+        echo "CHECKPOINT=$(cat CHECKPOINT-COMMIT)" >> $GITHUB_ENV
+    - name: Get kernel source at checkpoint
+      uses: libbpf/ci/get-linux-source@thefallentree
+      with:
+        repo: 'https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git'
+        rev: ${{ env.CHECKPOINT }}
+        dest: '${{ github.workspace }}/.kernel'
+    - name: Prepare to build BPF selftests
```
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/vmtest/action.yml:36]: yes still need to run make olddefconfig , but maybe not make prepare, but I found it doesn't harm anything

File: .github/actions/vmtest/action.yml [hunk 4]
```diff
@@ -5,31 +5,72 @@ inputs:
...
+        kernel: ${{ inputs.kernel }}
+    # 4. prepare rootfs
+    - name: prepare rootfs
+      uses: libbpf/ci/prepare-rootfs@thefallentree
+      with:
+        kernel: ${{ inputs.kernel }}
+        project-name: 'libbpf'
+        arch: ${{ env.ARCH }}
+    # 5. run selftest in QEMU
+    - name: Run selftests
```
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/vmtest/action.yml:77]: it's hard coded in prepare-rootfs steps.  that steps encode a setup command to one of the init steps

File: .github/actions/vmtest/action.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply .github/actions/vmtest/action.yml:52]: it slows down UI. Did selftests build ever got stuck? Did you ever have to use that output for anything? If there are errors, they will be output to stderr and we'll see them. So I still think we should add back /dev/null for stdout. Same also for kernel build log, it's not useful at all.

File: .github/actions/build-selftests/build_selftests.sh [hunk 5]
```diff
@@ -2,41 +2,41 @@
...
+cd -
```
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/build-selftests/build_selftests.sh:34]: ack

File: .github/actions/vmtest/action.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (anakryiko) [reply .github/actions/vmtest/action.yml:36]: Do you know why BPF selftest need make olddefconfig? Is it for test module or something else?

**Comment by thefallentree** [comment]
> > PTAL, I've cleaned up the output log
> 
> not really?.. first few steps are still messed up:

no, I left those output intentionally because it shows what kernel version etc we are testing against

**Comment by anakryiko** [comment]
> > > PTAL, I've cleaned up the output log
> > 
> > 
> > not really?.. first few steps are still messed up:
> 
> no, I left those output intentionally because it shows what kernel version etc we are testing against

Look at the very first screen shot above (current output). It has kernel version (LATEST), linux repo sha, etc. But also a clean grouping with no random unformatted output.

**Comment by thefallentree** [comment]
> > > > PTAL, I've cleaned up the output log
> > > 
> > > 
> > > not really?.. first few steps are still messed up:
> > 
> > 
> > no, I left those output intentionally because it shows what kernel version etc we are testing against
> 
> Look at the very first screen shot above (current output). It has kernel version (LATEST), linux repo sha, etc. But also a clean grouping with no random unformatted output.

its the same thing ,but now the extra few lines of outputs are  inserted by GitHub from running  external actions , we don’t have ways to control that.

**Comment by anakryiko** [comment]
> > Look at the very first screen shot above (current output). It has kernel version (LATEST), linux repo sha, etc. But also a clean grouping with no random unformatted output.
> 
> its the same thing ,but now the extra few lines of outputs are inserted by GitHub from running external actions , we don’t have ways to control that.

I see :( too bad, but ok then

Alright, so besides shorter log output from kernel build and selftests build, is there anything else left on this? Looks good to me otherwise.

File: .github/actions/vmtest/action.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/vmtest/action.yml:36]: yes, it is for the test module

File: .github/actions/vmtest/action.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (thefallentree) [reply .github/actions/vmtest/action.yml:52]: done.

**Review by anakryiko — APPROVED** [review]
LGTM, thanks. As agreed offline, let's wait a bit for kernel-patches/vmtest part to also be switched before landing this.

**Pull request merged**